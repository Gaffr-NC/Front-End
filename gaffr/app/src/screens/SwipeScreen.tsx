import React, { Component } from 'react';
import Swiper from 'react-native-deck-swiper';
import {
  StyleSheet,
  View,
  Text,
  Image,
  AsyncStorage,
  RefreshControl,
  ScrollView
} from 'react-native';
import { FontAwesome } from '@expo/vector-icons';
import {
  getUsers,
  addMatch,
  getUserById,
  getSuitableLandlords,
  capitalise
} from '../utils/index';
import { NavigationScreenProp } from 'react-navigation';
import { User } from '../utils/interfaces';

interface Props {
  navigation: NavigationScreenProp<any, any>;
}

export default class SwipeScreen extends Component<Props> {
  state = {
    uid: '',
    cards: [],
    swipedAllCards: false,
    swipeDirection: '',
    isSwipingBack: false,
    cardIndex: 0,
    refreshing: false
  };

  static navigationOptions: {
    // header: { visible: false };
    headerVisible: false;
    header: null;
  };

  componentDidMount = async () => {
    const uid = await AsyncStorage.getItem('uid');
    await this.setState({ uid });
    this.fetchData();
  };
  fetchData = async () => {
    const { uid } = this.state;
    if (uid) {
      const user = await getUserById(uid, 'tenants');
      const landlords = user.preferences
        ? await getSuitableLandlords(user.preferences, uid)
        : await getUsers('landlords');
      this.setState({
        uid,
        cards: landlords.filter((landlord: User) => landlord.property),
        cardIndex: 0,
        swipedAllCards: false
      });
    }
  };

  onRefresh = async () => {
    this.setState({ refreshing: true });
    await this.fetchData().then(() => {
      this.setState({ refreshing: false });
    });
  };

  renderCard = (cardData: any) => {
    return (
      cardData && (
        <View style={styles.swiperContainer}>
          <View style={styles.card}>
            <Image
              style={styles.image}
              source={{
                uri: cardData.property.images[0]
              }}
            />
            <View style={styles.textBox}>
              <Text style={styles.text}>
                <FontAwesome name="home" size={30} />{' '}
                {capitalise(cardData.property.propertyType)}
              </Text>
              <Text style={styles.text}>
                <FontAwesome name="bed" size={30} />{' '}
                {cardData.property.bedrooms}
              </Text>
              <Text style={styles.text}>
                <FontAwesome name="gbp" size={30} /> {cardData.property.price}
              </Text>
            </View>
          </View>
        </View>
      )
    );
  };

  onClickCard = (cardIndex: number) => {
    const card: User = this.state.cards[cardIndex];
    const { property } = card;
    if (property) {
      this.props.navigation.navigate('PropertyProfile', {
        images: property.images,
        area: property.city,
        price: property.price,
        propertyType: property.propertyType,
        description: property.description,
        bedrooms: property.bedrooms,
        smokingAllowed: property.smokingAllowed,
        petsAllowed: property.petsAllowed
      });
    }
  };

  onSwipedAllCards = () => {
    this.setState({
      swipedAllCards: true
    });
  };
  onSwipeRight = async (cardIndex: number) => {
    const { uid, cards } = this.state;
    const card: User = cards[cardIndex];
    if (card.id && uid) {
      addMatch(card.id, uid);
    }
  };

  render() {
    const { cards, swipedAllCards, refreshing } = this.state;
    return cards.length && !swipedAllCards ? (
      <ScrollView
        contentContainerStyle={{ flex: 1 }}
        refreshControl={
          <RefreshControl
            refreshing={this.state.refreshing}
            onRefresh={this.onRefresh}
          />
        }
      >
        {!refreshing && (
          <Swiper
            stackSize={6}
            backgroundColor={'#dcd1e8'}
            verticalSwipe={false}
            cards={this.state.cards}
            cardIndex={this.state.cardIndex}
            cardVerticalMargin={40}
            renderCard={this.renderCard}
            onSwipedAll={this.onSwipedAllCards}
            onSwipedRight={this.onSwipeRight}
            showSecondCard={true}
            overlayLabels={{
              left: {
                title: '✗',
                style: {
                  label: {
                    backgroundColor: 'transparent',
                    borderColor: '#f44242',
                    color: '#f44242',
                    borderWidth: 7
                  },
                  wrapper: {
                    flexDirection: 'column',
                    alignItems: 'flex-end',
                    justifyContent: 'flex-start',
                    marginTop: 30,
                    marginLeft: -30
                  }
                }
              },
              right: {
                title: '✓',
                style: {
                  label: {
                    backgroundColor: 'transparent',
                    borderColor: '#2fc47c',
                    color: '#2fc47c',
                    borderWidth: 7
                  },
                  wrapper: {
                    flexDirection: 'column',
                    alignItems: 'flex-start',
                    justifyContent: 'flex-start',
                    marginTop: 30,
                    marginLeft: 30
                  }
                }
              }
            }}
            animateOverlayLabelsOpacity
            animateCardOpacity
            onTapCard={(cardIndex: number) => this.onClickCard(cardIndex)}
          />
        )}
      </ScrollView>
    ) : (
      <ScrollView
        contentContainerStyle={{
          flex: 1,
          justifyContent: 'center',
          alignContent: 'center',
          backgroundColor: '#dcd1e8'
        }}
        refreshControl={
          <RefreshControl
            refreshing={this.state.refreshing}
            onRefresh={this.onRefresh}
          />
        }
      >
        <Text style={{ textAlign: 'center' }}>
          No more gaffs matching your criteria: pull down to refresh?
        </Text>
      </ScrollView>
    );
  }
}

const styles = StyleSheet.create({
  swiperContainer: {
    flex: 1
  },
  textBox: {
    flex: 0.5,
    justifyContent: 'flex-start',
    height: 50,
    alignItems: 'flex-start',
    borderRadius: 10
  },
  card: {
    flex: 1,
    backgroundColor: '#f9f4f5',
    padding: 10,
    color: '#000000',
    borderColor: '#502f4c',
    borderWidth: 3,
    borderRadius: 10
    // justifyContent: 'flex-start',
    // alignItems: 'center',
    // marginBottom: 20
  },
  text: {
    marginLeft: 20,
    fontSize: 20,
    backgroundColor: 'transparent',
    color: '#000000'
  },
  image: {
    flex: 1,
    borderRadius: 10
  }
});
