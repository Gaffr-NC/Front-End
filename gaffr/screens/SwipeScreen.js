import React, { Component } from "react";
import { View, Text } from "react-native";

export default class SwipeScreen extends Component {
  static navigationOptions = {
    title: "Swiper"
  };
  render() {
    return (
      <View style={{ flex: 1, alignItems: "center", justifyContent: "center" }}>
        <Text>Swipe Screen</Text>
      </View>
    );
  }
}
