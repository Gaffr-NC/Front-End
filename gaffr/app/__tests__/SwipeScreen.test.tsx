import React from 'react';
import renderer from 'react-test-renderer';
import SwipeScreen from '../src/screens/SwipeScreen';
let props: any;
it('SwipeScreen renders without crashing', () => {
  const rendered = renderer.create(<SwipeScreen {...props} />).toJSON();
  expect(rendered).toBeTruthy();
});
it('Tests SwipeScreen against snapshot', () => {
  const tree = renderer.create(<SwipeScreen {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});
