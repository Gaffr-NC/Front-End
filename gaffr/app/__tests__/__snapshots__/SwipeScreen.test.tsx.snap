// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests SwipeScreen against snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#f9f4f5",
      "color": "#f9f4f5",
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "backgroundColor": "#4FD0E9",
          "marginBottom": 0,
          "marginTop": 0,
        },
        Object {},
      ]
    }
  >
    <View
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "zIndex": -6,
          },
        ]
      }
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flex": 1,
          "height": 1174,
          "left": 20,
          "opacity": 1,
          "position": "absolute",
          "top": 80,
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
            Object {
              "rotate": "0deg",
            },
          ],
          "width": 710,
          "zIndex": 1,
        }
      }
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flex": 1,
          "height": 1174,
          "left": 20,
          "position": "absolute",
          "top": 80,
          "transform": Array [
            Object {
              "scale": 0.97,
            },
            Object {
              "translateY": 10,
            },
          ],
          "width": 710,
          "zIndex": -1,
        }
      }
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flex": 1,
          "height": 1174,
          "left": 20,
          "position": "absolute",
          "top": 80,
          "transform": Array [
            Object {
              "scale": 0.9400000000000001,
            },
            Object {
              "translateY": 20,
            },
          ],
          "width": 710,
          "zIndex": -2,
        }
      }
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flex": 1,
          "height": 1174,
          "left": 20,
          "position": "absolute",
          "top": 80,
          "transform": Array [
            Object {
              "scale": 0.91,
            },
            Object {
              "translateY": 30,
            },
          ],
          "width": 710,
          "zIndex": -3,
        }
      }
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flex": 1,
          "height": 1174,
          "left": 20,
          "position": "absolute",
          "top": 80,
          "transform": Array [
            Object {
              "scale": 0.88,
            },
            Object {
              "translateY": 40,
            },
          ],
          "width": 710,
          "zIndex": -4,
        }
      }
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flex": 1,
          "height": 1174,
          "left": 20,
          "position": "absolute",
          "top": 80,
          "transform": Array [
            Object {
              "scale": 0.85,
            },
            Object {
              "translateY": 50,
            },
          ],
          "width": 710,
          "zIndex": -5,
        }
      }
    />
  </View>
</View>
`;
